src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,414 @@
+};
+
+/// returns the appropriate `Halide::Type` for the given operation type
+Type amx_op_type_result_type(AMXOpType op_ty) {
+    switch (op_ty) {
+    case AMXOpType::Int8:
+        return Int(32, 256);
+    case AMXOpType::Bf16:
+        return Float(32, 256);
+    default:
+        return Type();
+    }
+}
+
+const auto wild_i32 = Variable::make(Int(32), "*");
+const auto wild_i32x = Variable::make(Int(32, 0), "*");
+
+Tile<2> is_2d_tile_index(const Expr &e) {
+    // ramp(ramp(base, 1, 4), x4(stride), 4)
+    vector<Expr> matches;
+    if (const auto *r1 = e.as<Ramp>()) {
+        if (const auto *r2 = r1->base.as<Ramp>()) {
+            auto ramp_2d_pattern = Ramp::make(Ramp::make(wild_i32, wild_i32, r2->lanes), Broadcast::make(wild_i32, r2->lanes), r1->lanes);
+            if (expr_match(ramp_2d_pattern, e, matches)) {
+                return {true, std::move(matches[0]), {std::move(matches[2]), std::move(matches[1])}, {r1->lanes, r2->lanes}};
+            }
+        }
+    }
+    return {};
+}
+
+Tile<3> is_3d_tile_index(const Expr &e) {
+    vector<Expr> matches;
+
+    // there could be a sub node
+    const Sub *sub = e.as<Sub>();
+
+    const Add *add = nullptr;
+
+    if (sub) {
+        add = sub->a.as<Add>();
+    } else {
+        add = e.as<Add>();
+    }
+
+    if (!add) {
+        return {};
+    }
+
+    const auto &first = add->a;
+    const auto &second = add->b;
+
+    // ramp(x[x*r](base), x[x*r](stride), x) + x[x*y](ramp(idx, 1, r))
+
+    const auto *r1 = first.as<Ramp>();
+    const auto *b2 = second.as<Broadcast>();
+    if (!r1 && !b2) {
+        // Try switching the order
+        r1 = second.as<Ramp>();
+        b2 = first.as<Broadcast>();
+    }
+    if (!r1 || !b2) {
+        return {};
+    }
+
+    const auto *b1 = r1->base.as<Broadcast>();
+    const auto *r2 = b2->value.as<Ramp>();
+
+    if (!b1 || !r2) {
+        return {};
+    }
+
+    int x_tile = r1->lanes;
+    int r_tile = r2->lanes;
+    int y_tile = b1->lanes / r_tile;
+    if (y_tile != b2->lanes / x_tile) {
+        return {};
+    }
+
+    auto pattern1 = Ramp::make(Broadcast::make(wild_i32, b1->lanes), Broadcast::make(wild_i32, b1->lanes), r1->lanes);
+    if (!expr_match(pattern1, first, matches)) {
+        return {};
+    }
+    Expr base = std::move(matches[0]);
+    Expr x_stride = std::move(matches[1]);
+
+    auto pattern2 = Broadcast::make(Ramp::make(wild_i32, wild_i32, r2->lanes), b2->lanes);
+    if (!expr_match(pattern2, second, matches)) {
+        return {};
+    }
+    base += std::move(matches[0]);
+    Expr r_stride = std::move(matches[1]);
+
+    if (sub) {
+        Expr adj = sub->b;
+        const Broadcast *bcast = adj.as<Broadcast>();
+
+        if (!bcast) {
+            return {};
+        }
+
+        if (bcast->lanes != b1->lanes * r1->lanes) {
+            return {};
+        }
+
+        base -= bcast->value;
+    }
+
+    return {true, base, {x_stride, 0, r_stride}, {x_tile, y_tile, r_tile}};
+}
+
+struct NewMatmul {
```

Comment:
Identifiers with "New" in the name make me nervous, since they usually get stale quickly and refer to a not-really-new thing. Is there a better name?